### PR TITLE
Added support for the entire AT24C device family

### PR DESCRIPTION
**Pull request by DoNotTapeOver**
## Description
* Added classes to support AT24C01C, AT24C02C, AT24C04C, AT24C08C, AT24C16C, AT24C32C, AT24C64C, AT24C128C, and AT24C256C ICs.
* Moved the existing AT24128C class into the AT24CXX directory to be consistent with other device family implementations in nanoFramework.Iot.Device.
* Changed the namespace of AT24128C from Iot.Device.At24C128C to Iot.Device.At24CXX.
* Changed the constructor of AT24128C to accept an I2cDevice reference instead of being passed address and i2cBus. The existing AT24128C implementation was not consistent with other device implementations in nanoFramework.Iot.Device.
* Added support for sequential read feature.

## Motivation and Context
The AT24 EEPROM family contains 9 different ICs each with a different memory size, page count, and page size. Each chip can only write up to its page size in bytes per write cycle. Using the appropriate class now throws an exception if more data is written that the chip will allow per write cycle. The AT24 family also contains an internal address counter that allows for sequentially reading bytes. The existing implementation did not provide a way to take advantage of this feature.

## How Has This Been Tested?
The provided sample code was tested using an ESP32 (ESP32-WROOM-32) and an AT24C32C.

## Types of changes

- [x] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [x] New feature (non-breaking change which adds functionality to code)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project (only if there are changes in source code).
- [x] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/master/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [ ] I have added new tests to cover my changes.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/nanoframework/nanoFramework.IoT.Device?pullRequest=512) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/nanoframework/nanoFramework.IoT.Device?pullRequest=512) before we can accept your contribution.<br/><br/>:x: DoNotTapeOver [sign now](https://cla.dotnetfoundation.org/nanoframework/nanoFramework.IoT.Device?pullRequest=512)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/nanoframework/nanoFramework.IoT.Device?pullRequest=512) it.</sub>

**Comment by Ellerbach** [comment]
Thanks a lot, a general comment: the naming convention goes for lower case "x" in the name when replacing a character. That applies to everything: the directory name, the file name, the class name, the namespaces.
Can you please adjust all this?
Note that on Windows, if you want to adjust a file name or a directory name, you have to:
* Rename to something else (add the number 2 for example).
* Do a commit
* Rename to the proper name you want
* Do a commit
* Make the rest of your changes

File: devices/At24CXX/AT24Base.cs [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+            byte[] writeBuffer = CreateWriteBuffer(address, data.Length);
+            data.CopyTo(writeBuffer, 2);
+
+            I2cTransferResult result = _i2cDevice.Write(writeBuffer);
+ 
+            // Account for memory address that was sent to device along with the data to be written
+            return result.BytesTransferred - 2;
+        }
+
+        private void Dispose(bool disposing)
```
Comment (Ellerbach) [inline devices/At24CXX/AT24Base.cs:187]: we don't really use this pattern anymore. You can move all the logic in the Dispose() function

File: devices/At24CXX/AT24C01C.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    /// I2C EEPROM with 1Kbit (128 bytes) of memory internally organized as 16 pages containing 8 bytes each.
+    /// </summary>
+    public sealed class AT24C01C : AT24Base
+    {
+        /// <summary>
+        /// Initializes a new instance of an AT24C01C device.
+        /// </summary>
+        /// <param name="i2cDevice">The I2C device to use for communication.</param>
+        public AT24C01C(I2cDevice i2cDevice)
+                : base(i2cDevice, 8, 16)
```
Comment (Ellerbach) [inline devices/At24CXX/AT24C01C.cs:18]: I think this one has too much spaces before the ': base'

File: devices/At24CXX/samples/At24CXX.samples.nfproj [hunk 3]
```diff
@@ -12,10 +12,10 @@
     <OutputType>Exe</OutputType>
     <AppDesignerFolder>Properties</AppDesignerFolder>
     <FileAlignment>512</FileAlignment>
-    <RootNamespace>Iot.Device.At24C128C.Samples</RootNamespace>
-    <AssemblyName>Iot.Device.At24C128C.Samples</AssemblyName>
+    <RootNamespace>Iot.Device.At24CXX.Samples</RootNamespace>
```
Comment (Ellerbach) [inline devices/At24CXX/samples/At24CXX.samples.nfproj:15]: You will have to adjust here as well for the xx. So should be At24Cxx

File: devices/At24C128C/version.json [hunk 4]
```diff
@@ -1,16 +0,0 @@
-{
```
Comment (Ellerbach) [inline devices/At24C128C/version.json:1]: you will need this file as well

File: devices/At24CXX/AT24Base.cs [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+namespace Iot.Device.At24CXX
+{
+    /// <summary>
+    /// Base class for common functionality of the AT24C serices of devices.
+    /// </summary>
+    public abstract class AT24Base : IDisposable
+    {
+        /// <summary>
+        /// Default I2C address for AT24Cx familly
```
Comment (Ellerbach) [inline devices/At24CXX/AT24Base.cs:15]: All comments should finish by a dot '.'

File: devices/At24CXX/AT24Base.cs [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+        /// </summary>
+        public const byte DefaultI2cAddress = 0x50; // 01010000
+
+        private bool _isDisposed;
+        private readonly I2cDevice _i2cDevice;
+
+        /// <summary>
+        /// Number of pages available on the device.
```
Comment (Ellerbach) [inline devices/At24CXX/AT24Base.cs:23]: All properties comments should start with 'Gets' if it's only a public get like here or 'Gets or sets' if both.

**Review by Ellerbach — COMMENTED** [review]
Thanks for this enhancement. Few comments.
Also, we are moving to use a StyleCop linter. You'll find it in add-linter branch. May I ask you to add it?
It's about to copy the setting.StyleCop file, add an entry in the nfprj and add a package. See the ones which have this.
The linter is not needed in the samples, just the main binding project.

File: devices/At24Cxx/At24Cxx.nfproj [hunk 7]
```diff
@@ -8,56 +8,45 @@
...
     <ProjectTypeGuids>{11A8DD76-328B-46DF-9F39-F559912D0360};{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}</ProjectTypeGuids>
-    <ProjectGuid>{BAFCA06F-E7CA-4476-BA1C-829F09589F83}</ProjectGuid>
+    <ProjectGuid>14e43e63-9b26-4a68-ac48-95211300906f</ProjectGuid>
     <OutputType>Library</OutputType>
     <AppDesignerFolder>Properties</AppDesignerFolder>
     <FileAlignment>512</FileAlignment>
-    <RootNamespace>Iot.Device.At24C128C</RootNamespace>
-    <AssemblyName>Iot.Device.At24C128C</AssemblyName>
+    <RootNamespace>At24Cxx</RootNamespace>
+    <AssemblyName>At24Cxx</AssemblyName>
```
Comment (Ellerbach) [inline devices/At24Cxx/At24Cxx.nfproj:16]: the assembly name should contain `Iot.Device.`, same for the root namespace

File: devices/At24C128C/At24C128C.nuspec [hunk 8]
```diff
@@ -1,37 +0,0 @@
-<?xml version="1.0" encoding="utf-8"?>
```
Comment (Ellerbach) [inline devices/At24C128C/At24C128C.nuspec:1]: you need as well this one, rename the file and what's inside.

**Review by Ellerbach — COMMENTED** [review]
Thanks! Few things missing, most important is all related to the nuget :-) so signing the assembly, the docs and the nuspec file.
And thanks as well for adding the linter!

File: devices/At24Cxx/At24Base.cs [hunk 9]
```diff
@@ -9,7 +9,7 @@ namespace Iot.Device.At24Cxx
     /// <summary>
     /// Base class for common functionality of the At24C serices of devices.
     /// </summary>
-    public abstract class At24Base : IDisposable
```
Comment (Ellerbach) [inline devices/At24Cxx/At24Base.cs]: just keep it anyway! Part of our pattern for the bindings.

**Review by Ellerbach — APPROVED** [review]
Thanks a lot! Looks great!

**Comment by Ellerbach** [comment]
Please resolve the conflicts so we can go for a merge.

**Comment by Ellerbach** [comment]
Note: to help resolve the conflict, you may have to pull the develop branch to your branch and then resolve them. If you need help, please let us know.

**Comment by Ellerbach** [comment]
Seems that there are some changes that should not be present here. If you pulled the dev branch, resolve the conflict and then commit everything and push. Otherwise, you'll get un sync with the dev branch.

**Comment by Ellerbach** [comment]
Hi, so maybe the best is just:
* Close this PR
* Copy your new files into a safe place. 
* Go back to the the develop branch, pull all the changes. 
* Then create a new branch
* Place back your new files, delete the ones which need to be deleted
* Go again for this PR
Sorry for the mess generated, we'll get there :-)

**Pull request closed without merging**